db/src/db.ts

```diff
@@ -91,4 +96,84 @@ export class DB {
       return next;
     });
   }
+
+  /**
+   * Find documents matching query.
+   * @param query - a query constructed with Q methods.
+   * @return - an array of documents
+   */
+  public async find(query: Q.Query): Promise<Document[]> {
+    const { filter, limit, skip, orderBy } = query.toJSON();
+    const results: Document[] = [];
+    await new Promise((resolve, reject) => {
+      const it = this.db.iterator({
+        keyAsBuffer: false,
+        valueAsBuffer: false,
+      });
+      const next = (err: any, key: string, rawValue: string) => {
+        if (err) {
+          return reject(err);
+        }
+        if (!key) {
+          // Iteration complete
+          return it.end(resolve);
+        }
+        const value = JSON.parse(rawValue);
+        if (match({ key, value }, filter)) {
+          results.push({ key, value });
+        }
+        return it.next(next);
+      };
+      it.next(next);
+    });
+    const sortedResults = orderBy ? sortWith(orderBy.map(buildComparator), results) : results;
+    return sortedResults.slice(skip, limit === undefined ? undefined : (skip || 0) + limit);
+  }
+}
+
+export function buildComparator([p, direction]: Q.Order) {
+  return direction === Q.ASC ? ascend(path(p)) : descend(path(p));
+}
+
+export function match(doc: Document, filter: Q.Filter): boolean {
+  switch (filter.operator) {
+    case 'and':
+      return filter.filters.every((f) => match(doc, f));
+    case 'or':
+      return filter.filters.some((f) => match(doc, f));
+    case 'not':
+      return !match(doc, filter.filter);
+  }
+
+  const value = path(filter.path, doc);
+
+  switch (filter.operator) {
+    case 'eq':
+      return value === filter.value;
+    case 'ne':
+      return value !== filter.value;
+    case 'gt':
+      return typeof value === typeof filter.value && value as any > filter.value;
+    case 'gte':
+      return typeof value === typeof filter.value && value as any >= filter.value;
+    case 'lt':
+      return typeof value === typeof filter.value && value as any < filter.value;
+    case 'lte':
+      return typeof value === typeof filter.value && value as any <= filter.value;
+    case 'exists':
+      return value !== undefined;
+    case 'isNull':
+      return value === null;
+    case 'matches':
+      if (typeof value !== 'string') {
+        return false;
+      }
+      const regexp = new RegExp(filter.pattern, filter.caseInsensitive ? 'i' : undefined);
+      return regexp.test(value);
+    case 'startsWith':
+      if (typeof value !== 'string') {
+        return false;
+      }
+      return value.startsWith(filter.value);
+  }
```

Comment:
Yes, but if the client passes a filter from a different version to the server it might happen.  (E.g. consider using a serialized filter that comes from the future, because it's from a newer package that uses a newer version.  Or _something_.)
Let's have an exception / error report when that happens, please.  (Could be at an earlier validation phase too, of course)